### ISSUE TITLE
[Bug] nn.RMSNorm produces wildly incorrect normalization constants for bfloat16 tensors with small variances

### ISSUE THREAD

**Issue by LotanGutman** [pytorch/pytorch#189581]
### 🐛 Describe the bug

When `torch.nn.RMSNorm` is applied to a `bfloat16` tensor with a small variance (e.g., standard deviation around `0.05`), the output diverges significantly (by ~1.7x) from both the mathematical definition of RMSNorm and PyTorch's own `float32` output. 
This appears to be caused by a severe precision loss or underflow during the internal sum-of-squares accumulation when the input is natively in `bfloat16`.
### Minimal Reproducible Example
```python
import torch
import torch.nn as nn

torch.manual_seed(42) # seed to get the same results as me (although results are consistant across a variety of seeds)

# Params
D = 32
B, H, N = 2, 4, 1

# Downcast a tensor with a very small variance in float32 to bfloat16
x_fp32 = torch.randn(B, H, N, D, dtype=torch.float32) * 0.058
x_bf16 = x_fp32.to(torch.bfloat16)
norm_layer = nn.RMSNorm(D, elementwise_affine=False)
# Native RMSNorm acting on bfloat16 directly (buggy)
out_pytorch_bf16 = norm_layer(x_bf16)
# Native RMSNorm acting on float32 (stable)
out_pytorch_fp32 = norm_layer(x_fp32)
# Manual RMSNorm implementation in float32
var_manual = (x_fp32 ** 2).mean(dim=-1, keepdim=True)
out_manual_fp32 = x_fp32 / torch.sqrt(var_manual + 1e-5)
print("PyTorch bfloat16 output [0,0,0,:4]:")
print(out_pytorch_bf16[0,0,0,:4].float().tolist())
print("\nPyTorch float32 output [0,0,0,:4]:")
print(out_pytorch_fp32[0,0,0,:4].tolist())
print("\nManual float32 output [0,0,0,:4]:")
print(out_manual_fp32[0,0,0,:4].tolist())
ratio = out_manual_fp32[0,0,0,:4] / out_pytorch_bf16[0,0,0,:4].float()
print(f"\nDiscrepancy Ratio (Manual fp32 / PyTorch bf16): {ratio.mean().item():.3f}x")
```
Output
```text
PyTorch bfloat16 output [0,0,0,:4]:
[1.03125, 0.796875, 0.48046875, -1.125]
PyTorch float32 output [0,0,0,:4]:
[1.771738052368164, 1.3675109148025513, 0.8281811475753784, -1.9359601736068726]
Manual float32 output [0,0,0,:4]:
[1.7695422172546387, 1.3658159971237183, 0.8271547555923462, -1.933560848236084]
Discrepancy Ratio (Manual fp32 / PyTorch bf16): 1.718x
```

### Versions

I used PyTorch: 2.6.0+cu124, on a RTX 4060 gpu.

Hopefully this will get resolved soon. Although the standard way is to use fp32 for RMSNorm regardless, it's still a bug (Took me a bunch of time to debug why my custom inference kernal mismatched naive pytorch by more than quantization noise😵‍💫).

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

**Comment by harshaljanjani** [comment]
Hi @LotanGutman
Running your script on current main, this should already be fixed; it was resolved in the releases after 2.6.0 by [#142848](https://github.com/pytorch/pytorch/pull/142848) and the follow-up [#147203](https://github.com/pytorch/pytorch/pull/147203). I've got `[1.773, 1.367, 0.828, -1.938]` with a ratio of 0.998x. Upgrading to ≥ 2.7 or a recent nightly should resolve it for you, could you try this?

**Comment by LotanGutman** [comment]
Hi @harshaljanjani ,

Thanks for taking a look! That makes perfect sense :)
I am currently running `2.6.0`, which explains why the bug for me. 

I'll keep a manual `.float()` cast in my code for now, and drop it once I upgrade to `2.7`. 

Appreciate the quick help! Closing this issue.
